smu/geometry/smu_molecule.py

```diff
@@ -182,11 +256,80 @@ def place_bonds(self, state):
       if btype:
         add_bond(a1, a2, btype, result)
 
+    return result
+
+  def place_bonds(self, state, matching_parameters: MatchingParameters) -> Optional[dataset_pb2.BondTopology]:
+    """Place bonds corresponding to `state`.
+
+    Args:
+      state: bonding pattern to be placed.
+      matching_parameters: optional settings
+    Returns:
+      If successful, a BondTopology
+    """
+    bt = self.place_bonds_inner(state)
+    if not bt:
+      return None
+
+    if matching_parameters.neutral_forms_during_bond_matching and self._contains_both_oxygen_and_nitrogen:
+      if not self.assign_charged_atoms(bt):
+        return None
+      # all bonds matched has already been checked.
+      return bt
+
     # Optionally check whether all bonds have been matched
+#   print("Bonds placed")
     if not self._must_match_all_bonds:
-      return result
+      return bt
 
+#   print(*zip(self._current_bonds_attached, self._max_bonds))
     if not np.array_equal(self._current_bonds_attached, self._max_bonds):
       return None
 
-    return result
+    return bt
+
+  def assign_charged_atoms(self, bt:dataset_pb2.BondTopology) -> bool:
+    """Assign (N, N+) and (O, O-) possibilities in `bt`.
+
+    bt must contain both N and O atoms.
+    Note that we assume _must_match_all_bonds, and return None if that cannot
+    be achieved.
+
+    Args:
+      bt: BondTopology, bt.atoms are updated in place
+    Returns: True if successful, False otherwise
+    """
+
+    carbon = dataset_pb2.BondTopology.ATOM_C
+    hydrogen = dataset_pb2.BondTopology.ATOM_H
+    fluorine = dataset_pb2.BondTopology.ATOM_F
+    nitrogen = dataset_pb2.BondTopology.ATOM_N
+    npos = dataset_pb2.BondTopology.ATOM_NPOS
+    oxygen = dataset_pb2.BondTopology.ATOM_O
+    oneg = dataset_pb2.BondTopology.ATOM_ONEG
+    net_charge = 0
+    for i, atom in enumerate(bt.atoms):
+      if atom in [carbon, hydrogen, fluorine]:
+        if self._max_bonds[i] != self._current_bonds_attached[i]:
+          return False
+      elif atom in [nitrogen, npos]:
+        if self._current_bonds_attached[i] == 4:
+          bt.atoms[i] = npos
+          net_charge += 1
+        elif self._current_bonds_attached[i] == 3:
+          bt.atoms[i] = nitrogen
+        else:
+          return False
+      elif atom in [oxygen, oneg]:
+        if self._current_bonds_attached[i] == 2:
+          bt.atoms[i] = oxygen
+        elif self._current_bonds_attached[i] == 1:
+          bt.atoms[i] = oneg
+          net_charge -= 1
+        else:    # Should never happen
+          return False
```

Comment:
An assert or an exception woudl be ideal for things that shoudl not happen.

Reply:
Actually when I put in an assert, I found that it was a case I had not covered - an Oxygen atom that was not attached. I had not thought of that, but of course unattached atoms happen here.